docs/RFCS/20220711_external_connections.md

```diff
@@ -0,0 +1,407 @@
+client interacting with the external connection. These query parameters often contain sensitive
+information such as access keys, certificates, usernames, and passwords, and depending on the
+provider, may have to be URL escaped or encoded to a particular format.
+
+Having to enter a multi-line, error-prone URI every time a user wants to interact with disaster
+recovery features or changefeeds, creates a poor user experience. Further, it is critically
+important from a security point of view to be able to delegate the creation, management, and usage
+of an external connection to different roles/users in the cluster. Our current implementation
+mandates that the user connecting to the external connection, for example, creating a changefeed,
+must have admin-level access to secret keys, certificates, and resource configuration parameters.
+This prevents the adoption of such database operations in environments where the admin-level
+operator is not the same as the database developer.
+
+The jobs that interact with external connections are usually long-running in nature. This means that
+they likely access the external connection more than once during their execution. To enable this,
+every job persists its URI in the `system.jobs` table. Every job is individually responsible for
+redacting sensitive information, before displaying the persisted URI via the various job
+observability tools that CockroachDB offers. There have been several occasions where we have leaked
+secrets because a job had forgotten to sanitize the URI before displaying it.
+
+At the root of these shortcomings is the tight coupling between the systems that use the external
+connection, and the representation of the external connection. Every operation that accesses an
+external connection is also made aware of the raw URI that is parsed to configure, authenticate, and
+interact with the connection. This RFC proposes separating the creation and management of an
+External Connection object from its use. This separation would facilitate the development of a
+standardized interface and a fine-grained permission model, which would create a more user-friendly,
+secure experience.
+
+## Technical design
+
+### Creating an External Connection
+
+An External Connection object represents a resource that resides outside of CockroachDB. In the
+scope of this RFC, the discussion is limited to the two types of external connections that are used
+by bulk operations and changefeeds, namely:
+
+- Storage: a connection capable of reading and writing bytes to an external store.
+
+- Key Management Service (KMS): a connection capable of interacting with keys hosted in a KMS, to
+  encrypt and decrypt data.
+
+External Connections will be persisted in a new system table called `system.external_connections`
+with the following schema:
+
+```sql
+CREATE TABLE system.external_connections (
+    name               STRING NOT NULL,
+    created            TIMESTAMPZ NOT NULL DEFAULT NOW(),
```

Comment:
Worth having an `updated` field since we are supporting `ALTER`?

Reply:
agreed, will add that in.